File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultRequirementsChecker.java
package co.infinum.princeofversions;

import android.os.Build;

import androidx.annotation.VisibleForTesting;

/**
 * Represent a concrete implementation of {@link RequirementChecker} that will be used by a default.
 */
class PrinceOfVersionsDefaultRequirementsChecker implements RequirementChecker {

    static final String KEY = "required_os_version";
    private final ApplicationVersionProvider provider;

    @VisibleForTesting
    PrinceOfVersionsDefaultRequirementsChecker(ApplicationVersionProvider provider) {
        this.provider = provider;
    }

    PrinceOfVersionsDefaultRequirementsChecker() {
        this(new ApplicationVersionProvider() {
            @Override
            public int provide() {
                return Build.VERSION.SDK_INT;
            }
        });
    }

    /**
     * Basic implementation of this method that is going to be used by default.
     * By default the only requirements we are checking, is if is required android lower than required.
     *
     * @param value Json data that contains all requirements for new update
     * @return true or false depending if are required requirements matched
     */
    @Override
    public boolean checkRequirements(String value) {
        int minSdk = Integer.parseInt(value);
        return minSdk <= provider.provide();
    }

    interface ApplicationVersionProvider {

        int provide();
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultRequirementsChecker.kt
package co.infinum.princeofversions

import android.os.Build
import androidx.annotation.VisibleForTesting

/**
 * The default [RequirementChecker] implementation.
 *
 * @param sdkVersionProvider A function that provides the current device's SDK version.
 * Defaults to returning `Build.VERSION.SDK_INT`.
 */
internal class PrinceOfVersionsDefaultRequirementsChecker @VisibleForTesting internal constructor(
    private val sdkVersionProvider: () -> Int
) : RequirementChecker {

    /**
     * Creates a requirement checker that uses the device's current SDK version.
     */
    constructor() : this({ Build.VERSION.SDK_INT })

    companion object {
        const val KEY = "required_os_version"
    }

    /**
     * Checks if the device's OS version meets the required minimum version.
     *
     * @param value The minimum required OS version as a String.
     * @return true if the device's OS version is greater than or equal to the required version.
     */
    override fun checkRequirements(value: String): Boolean {
        val minSdk = value.toInt()
        return minSdk <= sdkVersionProvider()
    }
}
